### ISSUE TITLE
Toil's Mesos executor doesn't handle temporary network partitions properly

### ISSUE THREAD

**Issue by joelarmstrong** [BD2KGenomics/toil#1553]
I had a worker node disconnect from, then reconnect to the Mesos master. All its jobs were marked as TASK_LOST, but they didn't seem to get shut down properly, and jobs that attempted to restart on that same node got this error:

```
f066954f-b182-47d2-8cb1-60c88b8ec238    Traceback (most recent call last):
f066954f-b182-47d2-8cb1-60c88b8ec238      File "/usr/local/lib/python2.7/dist-packages/toil/worker.py", line 336, in main
f066954f-b182-47d2-8cb1-60c88b8ec238        with fileStore.open(job):
f066954f-b182-47d2-8cb1-60c88b8ec238      File "/usr/lib/python2.7/contextlib.py", line 17, in __enter__
f066954f-b182-47d2-8cb1-60c88b8ec238        return self.gen.next()
f066954f-b182-47d2-8cb1-60c88b8ec238      File "/usr/local/lib/python2.7/dist-packages/toil/fileStore.py", line 488, in open
f066954f-b182-47d2-8cb1-60c88b8ec238        self.cleanCache(jobReqs)
f066954f-b182-47d2-8cb1-60c88b8ec238      File "/usr/local/lib/python2.7/dist-packages/toil/fileStore.py", line 1111, in cleanCache
f066954f-b182-47d2-8cb1-60c88b8ec238        assert self.jobID not in cacheInfo.jobState
f066954f-b182-47d2-8cb1-60c88b8ec238    AssertionError
```
presumably because the KtServerService was still running.

Excerpts:

Mesos master log at disconnection:
```
I0308 08:18:29.357434    11 master.cpp:1245] Agent f15072d4-b9e6-44df-a669-66e986da45f7-S205 at slave(1)@172.31.39.251:5051 (ip-172-31-39-251.us-west-2.compute.internal) disconnected
I0308 08:18:29.357456    11 master.cpp:2784] Disconnecting agent f15072d4-b9e6-44df-a669-66e986da45f7-S205 at slave(1)@172.31.39.251:5051 (ip-172-31-39-251.us-west-2.compute.internal)
I0308 08:18:29.357470    11 master.cpp:2803] Deactivating agent f15072d4-b9e6-44df-a669-66e986da45f7-S205 at slave(1)@172.31.39.251:5051 (ip-172-31-39-251.us-west-2.compute.internal)
I0308 08:18:29.357489    11 master.cpp:1264] Removing framework f15072d4-b9e6-44df-a669-66e986da45f7-0022 (toil) at scheduler-895ded3d-4d44-453e-a5ab-3d449034e41a@172.31.32.210:43711 from disconnected agent f15072d4-b9e6-44df-a669-66e986da45f7-S205 at slave(1)@172.31.39.251:5051 (ip-172-31-39-251.us-west-2.compute.internal) because the framework is not checkpointing
```

Mesos master log at reconnection:
```
I0308 08:18:31.647519    15 master.cpp:4713] Re-registering agent f15072d4-b9e6-44df-a669-66e986da45f7-S205 at slave(1)@172.31.39.251:5051 (ip-172-31-39-251.us-west-2.compute.internal)
I0308 08:18:31.647894    15 master.cpp:4940] Sending updated checkpointed resources  to agent f15072d4-b9e6-44df-a669-66e986da45f7-S205 at slave(1)@172.31.39.251:5051 (ip-172-31-39-251.us-west-2.compute.internal)
I0308 08:18:31.648150     8 hierarchical.cpp:559] Agent f15072d4-b9e6-44df-a669-66e986da45f7-S205 reactivated
I0308 08:18:31.648983    15 master.cpp:5002] Received update of agent f15072d4-b9e6-44df-a669-66e986da45f7-S205 at slave(1)@172.31.39.251:5051 (ip-172-31-39-251.us-west-2.compute.internal) with total oversubscribed resources 
```

Mesos worker log at disconnection/reconnection:
```
I0308 08:18:29.436141    27 status_update_manager.cpp:320] Received status update TASK_RUNNING (UUID: e570b0ce-8ca6-4132-a67a-f7106429d5f9) for task 1762 of framework f15072d4-b9e6-44df-a669-66e986da45f7-0022
I0308 08:18:29.436385    27 slave.cpp:3604] Forwarding the update TASK_RUNNING (UUID: e570b0ce-8ca6-4132-a67a-f7106429d5f9) for task 1762 of framework f15072d4-b9e6-44df-a669-66e986da45f7-0022 to master@172.31.32.210:5050
I0308 08:18:29.436482    27 slave.cpp:3514] Sending acknowledgement for status update TASK_RUNNING (UUID: e570b0ce-8ca6-4132-a67a-f7106429d5f9) for task 1762 of framework f15072d4-b9e6-44df-a669-66e986da45f7-0022 to executor(1)@172.31.39.251:38745
I0308 08:18:31.136001    22 slave.cpp:3694] Master marked the agent as disconnected but the agent considers itself registered! Forcing re-registration.
I0308 08:18:31.136268    22 slave.cpp:884] Re-detecting master
I0308 08:18:31.136304    22 slave.cpp:927] Detecting new master
I0308 08:18:31.136392    22 slave.cpp:895] New master detected at master@172.31.32.210:5050
I0308 08:18:31.136418    22 slave.cpp:916] No credentials provided. Attempting to register without authentication
I0308 08:18:31.136442    22 slave.cpp:927] Detecting new master
I0308 08:18:31.136611    36 status_update_manager.cpp:174] Pausing sending status updates
I0308 08:18:31.136641    36 status_update_manager.cpp:174] Pausing sending status updates
I0308 08:18:31.687386    23 slave.cpp:1197] Re-registered with master master@172.31.32.210:5050
I0308 08:18:31.687474    23 slave.cpp:1233] Forwarding total oversubscribed resources 
I0308 08:18:31.687611    23 slave.cpp:2387] Updating framework f15072d4-b9e6-44df-a669-66e986da45f7-0022 pid to scheduler-895ded3d-4d44-453e-a5ab-3d449034e41a@172.31.32.210:43711
I0308 08:18:31.687680    13 status_update_manager.cpp:181] Resuming sending status updates
W0308 08:18:31.687707    13 status_update_manager.cpp:188] Resending status update TASK_RUNNING (UUID: a36d3369-ed7e-4b7e-9bfe-2b9a9753fcec) for task 1758 of framework f15072d4-b9e6-44df-a669-66e986da45f7-0022
W0308 08:18:31.688012    13 status_update_manager.cpp:188] Resending status update TASK_RUNNING (UUID: 72842ca9-33e6-4102-84f7-fb0acc24006b) for task 1759 of framework f15072d4-b9e6-44df-a669-66e986da45f7-0022
W0308 08:18:31.688108    13 status_update_manager.cpp:188] Resending status update TASK_RUNNING (UUID: e570b0ce-8ca6-4132-a67a-f7106429d5f9) for task 1762 of framework f15072d4-b9e6-44df-a669-66e986da45f7-0022
I0308 08:18:31.688117     7 slave.cpp:3604] Forwarding the update TASK_RUNNING (UUID: a36d3369-ed7e-4b7e-9bfe-2b9a9753fcec) for task 1758 of framework f15072d4-b9e6-44df-a669-66e986da45f7-0022 to master@172.31.32.210:5050
```

The executor also had quite a few errors in its stderr:
```
Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/toil/fileStore.py", line 1468, in asyncWrite
    raise RuntimeError("The termination flag is set, exiting")
RuntimeError: The termination flag is set, exiting

Exception in thread Thread-5:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/usr/local/lib/python2.7/dist-packages/toil/fileStore.py", line 1468, in asyncWrite
    raise RuntimeError("The termination flag is set, exiting")
RuntimeError: The termination flag is set, exiting

Exception RuntimeError: RuntimeError('cannot join current thread',) in <bound method CachingFileStore.__del__ of <toil.fileStore.CachingFileStore object at 0x7f094b171e50>> ignored
```
but they didn't seem to outright crash the executor.

Full logs here:

[toil-master.txt.gz](https://github.com/BD2KGenomics/toil/files/828913/toil-master.txt.gz)
[mesos-master.INFO.gz](https://github.com/BD2KGenomics/toil/files/828878/mesos-master.INFO.gz)
[mesos-slave.INFO.gz](https://github.com/BD2KGenomics/toil/files/828876/mesos-slave.INFO.gz)
[worker-executor-logs.tar.gz](https://github.com/BD2KGenomics/toil/files/828877/worker-executor-logs.tar.gz)

**Comment by joelarmstrong** [comment]
Previous toil-master log was incomplete, here is the full one: [toil-master.txt.gz](https://github.com/BD2KGenomics/toil/files/828913/toil-master.txt.gz)

**Comment by cket** [comment]
I'm not seeing a way to trigger executor code when the slave is disconnected... A potential fallback could be having the leader process kill all jobs that have a TASK_LOST status prior to reissuing them.

**Comment by joelarmstrong** [comment]
Isn't it `reregistered` / `disconnected` here?:

https://github.com/BD2KGenomics/toil/blob/master/src/toil/batchSystems/mesos/executor.py#L67

**Comment by cket** [comment]
As far as I can tell those callbacks are triggered when the executor loses communication with the agent process, which isn't happening in this case (http://mesos.apache.org/documentation/latest/app-framework-development-guide/)

**Comment by joelarmstrong** [comment]
Oops, you're right. Hmm.

It looks like [this StackOverflow question(https://stackoverflow.com/questions/31689989/ensuring-that-a-mesos-task-is-not-running-after-a-task-lost-status-update) is only tangentially related (they are worried about a task running on the other side of the partition before it's healed, but here we have a task running even after the partition healed). But one of the answers seems to indicate that the Mesos master will direct the worker to kill the `TASK_LOST` tasks after the partition heals.

Maybe the task *was* killed after all, but the executor didn't clean up the cache properly? It seems like if the executor gets a `killTask` message, it just sends a SIGKILL and doesn't do any cleanup. Should the worker be resetting the cache info after a killed task?

**Comment by cket** [comment]
I think the task was still alive, at least when the 2nd job started. When a job starts it removes all old jobIDs from jobState if the old job's PID is dead. There is no need to reset the cache since subsequent jobs will clean it up themselves. 

So far the failover_timeout parameter seems promising

**Comment by cket** [comment]
Relevant documentation from http://mesos.apache.org/documentation/latest/high-availability-framework-guide/: "The agent is given a chance to reconnect until health checks timeout. If the agent does reconnect, any tasks for which TASK_LOST updates were previously sent will be killed." 

As far as I can tell this timeout isn't easily configurable. The health check documentation here https://github.com/apache/mesos/blob/master/docs/health-checks.md shows that it is only a small part of the health check mechanism, which we currently aren't using in Toil. We don't want to have to implement this whole feature just to configure the timeout, but it might be helpful anyway - we could potentially use this mechanism to detect & kill hanging jobs.

**Comment by joelarmstrong** [comment]
Sorry for the long delay in responding. Yes, looks like you're right and the agent is actually not instructed to kill the failing jobs if it reconnects within less than 10 mins (by default)--that really sucks. I've had some issues with this, especially since our jobs aren't really idempotent--if they fail, they can be retried, but they absolutely can't be run twice. Of course, that is at the root an issue on our end, but it would be nice to mitigate that somewhat.

Anecdotally, I've had luck with increasing the agent ping timeout (`--agent_ping_timeout`) to 60s, and letting the master miss 15 pings from the agent (`--max_agent_ping_timeouts`) before considering it disconnected. That gives it about 15 minutes to reconnect, I think, and the default is a bit more than a minute. I think `failover_timeout` has to do with failing over to a new Mesos master, when you're using Mesos with multiple masters (HA mode)--but maybe I'm misunderstanding the docs.

**Comment by joelarmstrong** [comment]
Closing this because this is Mesos's problem, not Toil's. The immediate problem was fixed in Mesos 1.0.1, but the underlying cause is probably still there.
